src/models/kinflated_gamma_poisson_model.py

```diff
@@ -0,0 +1,602 @@
+          a term for weighting the difference in 1+ reach.
+        """
+        # Estimate total number of potential impressions
+        Imax = N * dist.expected_value()
+        if Imax <= I:
+            return np.sum(np.array(h) ** 2)
+
+        freqs = N * dist.kreach(np.arange(1, len(h)), I / Imax)
+        kplus = N * dist.kplusreach(len(h), I / Imax)
+        hbar = np.array(list(freqs) + [kplus])
+
+        obj = np.sum((hbar - h) ** 2 / (hbar + 1e-6))
+
+        if np.isnan(obj):
+            logging.vlog(2, f"alpha {alpha} beta {beta} N {N} Imax {Imax}")
+            logging.vlog(2, f"h    {h}")
+            logging.vlog(2, f"hbar {hbar}")
+            raise RuntimeError("Invalid value of objective function")
+
+        return obj
+
+    def _exponential_poisson_reach(self, I: float, N: float, beta: float) -> float:
+        """Estimates reach in an exponential poisson model.
+
+        Estimates reach when I impressions are purchased from a population
+        of N users where the impression count of each user is distributed
+        according to the exponential poisson distribution with parameter beta.
+
+        Under the exponential poisson model, the probability that a
+        user will have n impressions is (1 - p) * p^{n-1}, where p =
+        beta/(beta + 1).  This can be worked out by using the
+        equivalence between the gamma poisson and the negative
+        binomial and setting alpha = r = 1.
+
+        The probability that a user is not reached, given that a fraction q
+        of available impressions is purchased, is given by
+           sum_{n=1}^{infinity} (1 - q)^n f(n),
+        where f(n) is the PMF of the above exponential poisson distribution.
+        This simplifies to (1 - p)(1 - q) / (1 - p(1 - q)).
+
+        The probability that a user is reached at least once is one minus
+        the above quantity.  Finally, the value of q is taken to be
+        I / ((beta + 1) * N), as the average number of impressions that a
+        user will have is beta + 1.
+
+        Args:
+          I:  The hypothetical number of impressions purchased.
+          N:  The audience size.
+          beta: The parameter of the exponential-poisson distribution.
+        Returns:
+          The expected reach.
+        """
+        return N * I * (beta + 1) / (N + beta * (N + I))
+
+    def _exponential_poisson_beta(self, I: float, N: float, R: float) -> float:
+        """Estimate beta given impressions, audience size and reach."""
+        if I * R - N * (I - R) <= 0:
+            return 1e3
+        return (N * (I - R)) / (I * R - N * (I - R))
+
+    def _exponential_poisson_N_from_beta(
+        self, I: float, R: float, beta: float
+    ) -> float:
+        """Estimate audience size from impressions, reach and beta."""
+        if I == R:
+            return 100.0 * R
+        return -(I * beta * R) / ((beta + 1) * (R - I))
+
+    def _exponential_poisson_N(self, I: float, R: float, mu_scale_factor=1.5) -> float:
```

Comment:
why using mu instead of directly using beta?

Reply:
In my view, beta is an internal model parameter that is hard to interpret, whereas mu is a quantity that is very easy to understand.  So, it seemed more natural to formulate this value in terms of mu.  I can change it to beta if you think that would be better.  Let me know.

Reply:
Makes sense to use mu here.